### DISCUSSION TITLE
why the compression ratio is different between command line and code？

### DISCUSSION THREAD

**Discussion by xiaodou1117** [lz4/lz4#1557] [Q&A] [answered]
when I use the code below to compress a 8kB file named 1.data：
```c++
LZ4F_preferences_t kPrefs0 = {};
    kPrefs0.frameInfo.blockMode = LZ4F_blockLinked;
    kPrefs0.frameInfo.blockSizeID = LZ4F_max256KB;
    size_t src_size = chunk->len();
    char* src_data = new char[src_size];
    memcpy(src_data, chunk->buf(), src_size);
    size_t max_dst_size = LZ4F_compressFrameBound(src_size,&kPrefs0);
    char* compressed_data = new char[max_dst_size];
    size_t compressed_data_size = LZ4F_compressFrame(compressed_data,max_dst_size,src_data,src_size,&kPrefs0);
```
   and the command line to compress the same file
   `lz4 1.data`
why their compression ratio is different？
the code got 1.8 ratio and the command line got 4.2 ratio

**Comment by t-mat** [comment]
Hi @xiaodou1117 

I think you miss something.  Technically, `lz4` may need `--no-crc`.  But your snippet should output perfectly same file which is produced by `lz4` CLI.

Please check the following trivial and working example code which is almost identical to your snippet.

https://gist.github.com/t-mat/6b3d55a9287333bd8bbf466fc9a065d5

and its experiment :

```sh
cd
git clone https://github.com/lz4/lz4.git lz4-1557
cd lz4-1557
git checkout -b v1.10.0

make

./programs/lz4 --no-crc --force README.md
# Compressed filename will be : README.md.lz4
# Compressed 5397 bytes into 3528 bytes ==> 65.37%

curl -JOL https://gist.githubusercontent.com/t-mat/6b3d55a9287333bd8bbf466fc9a065d5/raw/lz4-1557.c
cc ./lz4-1557.c -I lib -L lib -llz4 && ./a.out
# size of input [README.md] is 5397
# size of output [README.md.lz4-1557.lz4] is 3528

cmp --verbose README.md.lz4 README.md.lz4-1557.lz4
# (no difference)
```

**Reply by xiaodou1117** [reply, marked as answer]
thanks for answering. The code works well.
I found I miss the difference of binary file and simple file.
The code compress the binary form, and I copy it into a file then use CLI to compress it. So their ratio is different.
anyway, thank you.
